### ISSUE TITLE
Question - multiple invoke

### ISSUE THREAD

**Issue by dakom** [davidkpiano/xstate#321]
Actions and Activities can be launched in a sequence, for example specifying ['doThing1', 'doThing2'] as a state's actions.

However, it isn't clear how or if `invoke()` can be called multiple times in a state, for example to set different Promises off when the state is entered.

**Comment by davidkpiano** [comment]
You can invoke multiple services by specifying them as an array:

```js
{
  invoke: [
    { id: 'svc1', src: ... },
    { id: 'svc2', src: ... }
  ]
}
```

**Comment by davidkpiano** [comment]
https://xstate.js.org/docs/guides/communication.html#multiple-services

**Comment by RainerAtSpirit** [comment]
This is a followup question so I add that here. Is there a way to wait till all invoked services have reached their final state before moving on to the next state?
In the example below I'd like all childMachines be in their final state before moving on to the `success` state. Putting target: "success" into the onDone configuration doesn't do the job as the first response will move to `success` ignoring all other responses that are still in flight.

```typescript
states: {
    pending: {
      invoke: [
        {
          id: "topTenProjects",
          src: topTenProjects,
          onDone: {
            // target: "success",
            actions: assign((ctx, event) =>
              produce(ctx, draft => {
                draft.topTenProjects = event.data
              })
            )
          }
        },
        {
          id: "fundingTypes",
          src: fundingTypes,
          onDone: {
            // target: "success",
            actions: assign((ctx, event) =>
              produce(ctx, draft => {
                draft.fundingTypes = event.data
              })
            )
          }
        }
      ]
    },
    success: {}
  }
```

**Comment by davidkpiano** [comment]
@RainerAtSpirit That would probably make a good new issue. I haven't seen anything in SCXML that would notify "when all invoked services are done" although you can accomplish this with "when all parallel states are done" - https://xstate.js.org/docs/guides/final.html#parallel-states

That might be a bit verbose, but it's for good reason - parallel states can have invoked services on each of them, but they can also be normal states without invoked services.

You can create a helper function that cuts down on the extra config, something like:

```js

states: {
  pending: {
    ...parallelServices(svc1, svc2, svc3),
    onDone: { /* ... */ }
  }
}

```

**Comment by ilyavf** [comment]
Wow, this is really verbose. Here is how `parallelServices` helper may look like if we need to run different actions per service:
```js
function parallelServices(services) {
  return {
    type: 'parallel',
    states: services.reduce((acc, {src, actions}) => {
      return {
        ...acc,
        [src]: {
          initial: 'run',
          states: {
            run: {
              invoke: {
                src,
                onDone: {
                  actions,
                  target: 'done'
                }
              }
            },
            done: {
              type: 'final'    // <<< To get the parent state `onDone` we need to land into a `final` state per service.
            }
          }
        } 
      }
    }, {})
  }
}
```

And to use it:
```js
states: {
  pending: {
    ...parallelServices([{
        src: 'service1,
        actions: 'action1'
    }, {
        src: 'service2,
        actions: 'action2'
    }]),
    onDone: { /* ... */ }
  }
}
```

**Comment by BrunoQuaresma** [comment]
I tried to use it but I'm getting a TS error, so probably I'm doing something wrong:

```
  states: {
    initializing: {
      always: {
        target: "ready",
        invoke: [
          { id: "loadHighlightEvents", src: "loadHighlightEvents" },
          { id: "loadSelectionEvents", src: "loadSelectionEvents" },
          { id: "loadTextEditable", src: "loadTextEditable" },
        ],
      },
    },
```

I tried to use the `onDone` inside of the `always` but looks like it is not supported. The idea is to add some events to the DOM in parallel during the initialization.

**Comment by bzbetty** [comment]
old thread, but i had a bit of luck with

```
states: {
  pending: {
   invoke: [
       { src: 'service1', actions:... },
      { src: 'service2', actions:... }
    ],
    always: { target: 'X', guard: 'isLoaded' }
  }
}
```

where isLoaded checks context which is set by the actions of the invoke.
